modules/core/tests/client/components/LanguageSwitch.test.js

```diff
@@ -0,0 +1,70 @@
+import React from 'react';
+import { render, fireEvent } from '@testing-library/react';
+import '@testing-library/jest-dom/extend-expect';
+import i18n from '@/config/client/i18n';
+import locales from '@/config/shared/locales';
+import LanguageSwitch from '@/modules/core/client/components/LanguageSwitch.component';
+
+import * as users from '@/modules/users/client/api/users.api';
+const api = { users };
+
+const localeNames = locales.map(locale => locale.label);
+
+jest.mock('@/modules/users/client/api/users.api');
+
+beforeEach(() => i18n.changeLanguage('en'));
+
+describe('default presentation', () => {
+
+  it('has a menu item for each language', async () => {
+    const { getAllByRole } = render(<LanguageSwitch/>);
+    const items = getAllByRole('menuitem');
+    expect(items).toHaveLength(3);
+    const names = items.map(item => item.innerHTML);
+    expect(names).toEqual(localeNames);
+  });
+
+  it('can change language', async () => {
+    const { queryByRole, getByText } = render(<LanguageSwitch/>);
+    expect(queryByRole('button')).toHaveTextContent('English');
+    fireEvent.click(getByText('česky'));
+    expect(queryByRole('button')).toHaveTextContent('česky');
+  });
+
+  it('can save the language to the API', async () => {
+    const { getByText } = render(<LanguageSwitch saveToAPI={true}/>);
+    fireEvent.click(getByText('česky'));
+    expect(api.users.update).toHaveBeenCalledWith({ locale: 'cs' });
+  });
+
+});
+
+describe('select presentation', () => {
+
+  it('has an option for each language', async () => {
+    const { getAllByRole } = render(<LanguageSwitch presentation="select"/>);
+    const items = getAllByRole('option');
+    expect(items).toHaveLength(3);
+    const names = items.map(item => item.innerHTML);
+    expect(names).toEqual(localeNames);
+  });
+
+  it('can change language', async () => {
+    const { getByRole, getByText } = render(<LanguageSwitch presentation="select"/>);
+    expect(selectedOption(getByRole('combobox'))).toHaveTextContent('English');
+    fireEvent.change(getByRole('combobox'), { target: getByText('česky') });
+    expect(selectedOption(getByRole('combobox'))).toHaveTextContent('česky');
+  });
+
+  it('can save the language to the API', async () => {
```

Comment:
This test fails when run on its own. (`it` => `it.only` or `fit`)

My debugging suggests the following:
1. The mocked `api.users.update` still remembers having been called with `{ locale: 'cs' }` earlier and that's why it passes when the other test has run.
2. The line 62 (`fireEvent.change(getByRole('combobox'), { target: getByText('česky') });`) does change the select option, but fails to run the `select`'s `onChange` event (which is responsible for calling the api).

I'm going to sleep and look into this tomorrow (or asap later). To the points above:
1. First of all we need to ensure the test fails consistently, i.e. when called together and separately: I think we may be able e.g. to mock only the function `users.update` `beforeEach` and then [restore](https://jestjs.io/docs/en/mock-function-api#mockfnmockrestore) it in `afterEach`. Not sure what's the exact syntax. I've just found it in the jest docs.
2. How to fire the event on `select` so the `onChange` gets fired, I haven't found.

Does this make sense? Is it reproducible?

Reply:
Ah, good catch! Because the first save to API test makes the call, the mock check passes...

Actually, the error is quite simple, the save to API test for the select is missing the `saveToAPI={true}` option... which is why it never saves.

It's a good idea to reset the mocks too... I'll fix it up tomorrow, unless you want to before then. Gotta sleep now.

Reply:
The `saveToAPI={true}` was necessary but not sufficient. Additionally, `fireEvent.change(getByRole('combobox'), { target: { value: 'cs' } });` did the trick for me.

Reply:
Ah, indeed! I fixed it up now, I used:

```js
fireEvent.change(getByRole('combobox'), { target: { value: getByText('česky').value } });
```

... so we are still choosing the value based on the human readable label. To me, that code looks like it should work as:

```js
fireEvent.change(getByRole('combobox'), { target: getByText('česky') });
```

... basically, an object with a value property, but it doesn't work that way.

Select change events seem a bit of a weak point for this testing-library (I read though [this issue](https://github.com/testing-library/react-testing-library/issues/322) when I was originally writing the test, but still got it wrong).